File: DIP/DIP_010/C++/DIP_010_circle_detection_2.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace cv;
using namespace std;

int main()
{
    // Loading the actual image
    Mat image, gray;
    image=imread("CirclesDetection.jpg", IMREAD_COLOR);
    cv::imshow("Original image", image);
    cv::waitKey();

    // Edge cases
    if(image.empty()){
    	cout << "Error loading image" << endl;
    	return -1;
    }

    // converting the color image into grayscale
    cvtColor(image, gray, COLOR_BGR2GRAY);

    // Smoothing the image to get rid of noise, so no false detections
    // Here we used a 11x11 filter
    blur( gray, gray, Size(7,7), Point(-1,-1) );
    cv::imshow("Blurred_image", gray);
    //cv::imwrite("Blurred_image.jpg", gray);
    cv::waitKey();

    // Store the detected circles in a 3d-vector
    vector<Vec3f> circles;

    for( int maxR = 0; maxR < 200; maxR = maxR + 10)
    {

        // Apply hough transform
        HoughCircles(gray, circles, HOUGH_GRADIENT, 1,
                     100, // detect circle with different distance
                     100, 90,
                     0,maxR);

        // Draw the detected circles
        for( size_t i = 0; i < circles.size(); i++ )
        {
             Point center(cvRound(circles[i][0]), cvRound(circles[i][1]));
             int radius = cvRound(circles[i][2]);
             // Draw the center point of the circle
             circle( image, center, 3, Scalar(0,255,0), -1, 8, 0 );
        }
    }
    // Displaying the result
    cv::imshow("circles", image );
    cv::imwrite("output.jpg", image);
    cv::waitKey(0);
    return 0;
}

File: DIP/DIP_010/Python/DIP_010_circle_detection_2.py
import numpy as np
import cv2

image = cv2.imread("CirclesDetection.jpg",0)
output = cv2.imread("CirclesDetection.jpg",1)
cv2.imshow("Original image", image)
cv2.waitKey()

blurred = cv2.GaussianBlur(image,(11,11),0)

cv2.imshow("Blurred image", blurred)
cv2.waitKey()

#Finds circles in a grayscale image using the Hough transform
circles = cv2.HoughCircles(blurred, cv2.HOUGH_GRADIENT, 1, 100,
                             param1=100,param2=90,minRadius=0,maxRadius=200)

# cv2.HoughCircles function has a lot of parameters, so you can find more about it in documentation
# or you can use cv2.HoughCircles? in jupyter nootebook to get that 

# Check to see if there is any detection
if circles is not None:
    # If there are some detections, convert radius and x,y(center) coordinates to integer
    circles = np.round(circles[0, :]).astype("int")

    for (x, y, r) in circles:
        # Draw the circle in the output image
        cv2.circle(output, (x, y), r, (0,255,0), 3)
        # Draw a rectangle(center) in the output image
        cv2.rectangle(output, (x - 2, y - 2), (x + 2, y + 2), (0,255,0), -1)

cv2.imshow("Detections",output)
cv2.imwrite("output.jpg",output)
cv2.waitKey()